src/libsystemd/sd-bus/sd-bus.c

```diff
@@ -1377,38 +1377,61 @@ _public_ int sd_bus_open_user(sd_bus **ret) {
 
 int bus_set_address_system_remote(sd_bus *b, const char *host) {
         _cleanup_free_ char *e = NULL;
-        char *m = NULL, *c = NULL;
+        char *m = NULL, *c = NULL, *rbracket = NULL, *p = NULL;
 
         assert(b);
         assert(host);
 
-        /* Let's see if we shall enter some container */
-        m = strchr(host, ':');
+        /* Skip ":"s in ipv6 addresses */
+        if (*host == '[') {
+                char *t;
+
+                rbracket = strchr(host, ']');
+                if (!rbracket)
+                        return -EINVAL;
+                t = strndupa(host + 1, rbracket - host - 1);
+                e = bus_address_escape(t);
+                if (!e)
+                        return -ENOMEM;
+        }
+
+        /* Let's see if a port was given */
+        m = strchr(rbracket ? rbracket + 1 : host, ':');
+        if (m) {
+                char *t;
+
+                p = m + 1;
+
+                t = strchr(p, '/');
+                if (t)
+                        p = strndupa(p, t - p);
+
+                if (!in_charset(p, "0123456789"))
+                        return -EINVAL;
+        }
+
+        /* Let's see if a machine was given */
+        m = strchr(rbracket ? rbracket + 1 : host, '/');
         if (m) {
                 m++;
 
                 /* Let's make sure this is not a port of some kind,
                  * and is a valid machine name. */
-                if (!in_charset(m, "0123456789") && machine_name_is_valid(m)) {
-                        char *t;
-
-                        /* Cut out the host part */
-                        t = strndupa(host, m - host - 1);
-                        e = bus_address_escape(t);
-                        if (!e)
-                                return -ENOMEM;
-
-                        c = strjoina(",argv5=--machine=", m);
-                }
+                if (!in_charset(m, "0123456789") && machine_name_is_valid(m))
+                        c = strjoina(",argv7=--machine=", m);
         }
 
         if (!e) {
-                e = bus_address_escape(host);
+                char *t;
+
+                t = strndupa(host, strcspn(host, ":/"));
+
+                e = bus_address_escape(t);
                 if (!e)
                         return -ENOMEM;
         }
 
-        b->address = strjoin("unixexec:path=ssh,argv1=-xT,argv2=--,argv3=", e, ",argv4=systemd-stdio-bridge", c);
+        b->address = strjoin("unixexec:path=ssh,argv1=-xT,argv2=-p,argv3=", p ?: "22", ",argv4=--,argv5=", e, ",argv6=systemd-stdio-bridge", c);
```

Comment:
insert -p 22 here is not acceptable. A server I regularly connect to has ssh listening on some high port, and to make things easy i added a local ssh config snippet for it that ensures the right port is used for it. If you now hardcode port 22 here, you break things for me, as the command line override the config snippet...

Reply:
oh ok, I forgot that ssh can do that